src/cocotb_tools/runner.py

```diff
@@ -427,7 +427,14 @@ def test(
         self.env["COCOTB_RESULTS_FILE"] = str(results_xml_file)
 
         cmds: Sequence[_Command] = self._test_command()
-        self._execute(cmds, cwd=self.test_dir)
+        try:
+            self._execute(cmds, cwd=self.test_dir)
+        except subprocess.CalledProcessError as e:
+            if isinstance(self, Verilator):
+                # Verilator aborts on $fatal, but will still generate results.
+                self.log.info("Simulation failed: %d", e.returncode)
+            else:
+                raise
```

Comment:
I think what we need to do here is save the return code and set the exit code correctly after checking the results file.

So:
* Catch error here, log about simulation failure
* If results file does not exist, log about file not existing, `sys.exit()` with the return code from the simulator
* If results file exists and there are test failure, log about test failures, `sys.exit(1)`
* If results file exists and there are no test failures, `sys.exit()` with the return code from the simulator

Reply:
And we can leave the Makefile ignoring the error altogether the default as its what other simulators do.